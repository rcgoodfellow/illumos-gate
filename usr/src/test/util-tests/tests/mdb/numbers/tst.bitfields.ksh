#!/usr/bin/ksh
#
#
# This file and its contents are supplied under the terms of the
# Common Development and Distribution License ("CDDL"), version 1.0.
# You may only use this file in accordance with the terms of version
# 1.0 of the CDDL.
#
# A full copy of the text of the CDDL should have accompanied this
# source.  A copy of the CDDL is also available via the Internet at
# http://www.illumos.org/license/CDDL.
#

#
# Copyright 2022 Oxide Computer Company
#

#
# Sanity check parts of bitfields.
#

set -o pipefail

tst_root="$(dirname $0)/.."
tst_prog="$tst_root/progs/bitfields"
tst_outfile="/tmp/mdb.bitfield.out.$$"
tst_exp="$0.out"

#
# Top level ::print
#
$MDB -e "first::print -t broken_t" $tst_prog > $ODIR/stdout
$MDB -e "second::print -t broken6491_t" $tst_prog >> $ODIR/stdout

#
# ::print of specific members
#
$MDB -e "first::print broken_t brk_a" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_b" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_c" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_d" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_e" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_f" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_g" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_h" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_i" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_j" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_k" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_l" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_m" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t a" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t b" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t c" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t d" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t e" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t f" $tst_prog >> $ODIR/stdout

#
# ::printf of members. Note, if ::printf said '%x\n' below then we would
# include the string "\n" (not a newline) in the output. Instead we rely
# upon the implicit newline from mdb -e.
#
$MDB -e "first::printf '%x' broken_t brk_a" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_b" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_c" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_d" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_e" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_f" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_g" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_h" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_i" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_j" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_k" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_l" $tst_prog >> $ODIR/stdout
$MDB -e "first::printf '%x' broken_t brk_m" $tst_prog >> $ODIR/stdout
$MDB -e "second::printf '%x' broken6491_t a" $tst_prog >> $ODIR/stdout
$MDB -e "second::printf '%x' broken6491_t b" $tst_prog >> $ODIR/stdout
$MDB -e "second::printf '%x' broken6491_t c" $tst_prog >> $ODIR/stdout
$MDB -e "second::printf '%x' broken6491_t d" $tst_prog >> $ODIR/stdout
$MDB -e "second::printf '%x' broken6491_t e" $tst_prog >> $ODIR/stdout
$MDB -e "second::printf '%x' broken6491_t f" $tst_prog >> $ODIR/stdout

#
# If we pipe the output of ::print that is a different bitfield logic
# path. So we take that all to a `::eval '.=K'` as a basic way to get it
# out.
#
$MDB -e "first::print broken_t brk_a | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_b | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_c | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_d | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_e | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_f | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_g | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_h | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_i | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_j | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_k | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_l | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "first::print broken_t brk_m | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t a | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t b | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t c | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t d | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t e | ::eval '.=K'" $tst_prog >> $ODIR/stdout
$MDB -e "second::print broken6491_t f | ::eval '.=K'" $tst_prog >> $ODIR/stdout
