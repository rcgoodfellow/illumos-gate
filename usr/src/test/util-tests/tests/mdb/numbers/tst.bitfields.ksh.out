broken_t {
    unsigned int:3 brk_a :3 = 0x3
    unsigned int:2 brk_b :2 = 0x3
    unsigned int:1 brk_c :1 = 0
    unsigned int:1 brk_d :1 = 0x1
    unsigned int:1 brk_e :1 = 0x1
    unsigned int:1 brk_f :1 = 0x1
    unsigned int:3 brk_g :3 = 0x3
    unsigned int:3 brk_h :3 = 0x5
    unsigned int:5 brk_i :5 = 0x3
    unsigned int:4 brk_j :4 = 0x9
    unsigned int:6 brk_k :6 = 0x13
    unsigned int:1 brk_l :1 = 0
    unsigned int:1 brk_m :1 = 0x1
}
broken6491_t {
    unsigned short:1 a :1 = 0x1
    unsigned short:8 b :8 = 0x2
    unsigned short:3 c :3 = 0
    unsigned short:2 d :2 = 0
    unsigned short:1 e :1 = 0x1
    unsigned short:1 f :1 = 0
}
brk_a = 0x3
brk_b = 0x3
brk_c = 0
brk_d = 0x1
brk_e = 0x1
brk_f = 0x1
brk_g = 0x3
brk_h = 0x5
brk_i = 0x3
brk_j = 0x9
brk_k = 0x13
brk_l = 0
brk_m = 0x1
a = 0x1
b = 0x2
c = 0
d = 0
e = 0x1
f = 0
3
3
0
1
1
1
3
5
3
9
13
0
1
1
2
0
0
1
0
                3               
                3               
                0               
                1               
                1               
                1               
                3               
                5               
                3               
                9               
                13              
                0               
                1               
                1               
                2               
                0               
                0               
                1               
                0               
